stdlib/messaging/kafka/src/main/java/org/ballerinalang/messaging/kafka/impl/KafkaRecordConsumer.java

```diff
@@ -0,0 +1,177 @@
+
+    private static final Logger logger = LoggerFactory.getLogger(KafkaRecordConsumer.class);
+
+    private KafkaConsumer<byte[], byte[]> kafkaConsumer;
+    private Duration pollingTimeout = Duration.ofMillis(1000);
+    private int pollingInterval = 1000;
+    private boolean decoupleProcessing = true;
+    private String groupId;
+    private KafkaListener kafkaListener;
+    private String serviceId;
+    private int consumerId;
+    private ScheduledExecutorService executorService = Executors.newScheduledThreadPool(1);
+    private ScheduledFuture pollTaskFuture;
+
+    public KafkaRecordConsumer(KafkaListener kafkaListener,
+                               Properties configParams,
+                               String serviceId,
+                               int consumerId,
+                               KafkaConsumer<byte[], byte[]> kafkaConsumer) {
+        this.serviceId = serviceId;
+        this.consumerId = consumerId;
+        // Initialize Kafka Consumer.
+        if (Objects.isNull(kafkaConsumer)) {
+            this.kafkaConsumer = new KafkaConsumer<>(configParams);
+        } else {
+            this.kafkaConsumer = kafkaConsumer;
+        }
+        List<String> topics = (ArrayList<String>) configParams.get(KafkaConstants.ALIAS_TOPICS);
+        // Subscribe Kafka Consumer to given topics.
+        this.kafkaConsumer.subscribe(topics);
+        this.kafkaListener = kafkaListener;
+        if (configParams.get(KafkaConstants.ALIAS_POLLING_TIMEOUT) != null) {
+            this.pollingTimeout = Duration.ofMillis((Integer) configParams.get(KafkaConstants.ALIAS_POLLING_TIMEOUT));
+        }
+        if (configParams.get(KafkaConstants.ALIAS_POLLING_INTERVAL) != null) {
+            this.pollingInterval = (Integer) configParams.get(KafkaConstants.ALIAS_POLLING_INTERVAL);
+        }
+        if (configParams.get(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG) != null) {
+            this.decoupleProcessing = (Boolean) configParams.get(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG);
+        }
+        // This is to override default decouple processing setting if required.
+        if (configParams.get(KafkaConstants.ALIAS_DECOUPLE_PROCESSING) != null) {
+            this.decoupleProcessing = (Boolean) configParams.get(KafkaConstants.ALIAS_DECOUPLE_PROCESSING);
+        }
+        this.groupId = (String) configParams.get(ConsumerConfig.GROUP_ID_CONFIG);
+    }
+
+    private void poll() {
+        try {
+            ConsumerRecords<byte[], byte[]> recordsRetrieved = null;
+            try {
+                // Make thread-safe as kafka does not support multiple thread access
+                if (!closed.get()) {
+                    recordsRetrieved = this.kafkaConsumer.poll(this.pollingTimeout);
+                }
+            } catch (WakeupException e) {
+                // Ignore exception if connection is closing.
+                if (!closed.get()) {
+                    throw e;
```

Comment:
Is this getting caught at an upper layer and thrown out/returned as an `ErrorValue`?

Reply:
This should be ignored, as we are closing the connection. It is the recommended way to handle this.